README.md
# 🔥 NCPI FHIR Service

<p align="center">
  <a href="https://circleci.com/gh/ncpi-fhir/ncpi-api-fhir-service"><img src="https://img.shields.io/circleci/project/github/ncpi-fhir/ncpi-api-fhir-service.svg?style=for-the-badge"></a>
</p>

FHIR data service for NCPI uses the [Smile CDR FHIR server](https://smilecdr.com/docs/).

## Version

Smile CDR 2021.02.R05 (Odyssey) (See [changelogs](https://smilecdr.com/docs/introduction/changelog.html))

## Quickstart

NCPI FHIR services have been deployed into the three standard environments within the NCPI AWS account: development (DEV), QA, and production (PRD). The FHIR endpoints for each of these environments are:

- Development: https://ncpi-api-fhir-service-dev.kidsfirstdrc.org
- QA: https://ncpi-api-fhir-service-qa.kidsfirstdrc.org
- Production: https://ncpi-api-fhir-service.kidsfirstdrc.org

## 🚧 Server Access 🚧

**Note: We recognize that the process for accessing the deployed servers is not very user/developer friendly. This is temporary as we are still working on the NCPI infrastructure and server authentication. Please bear with us!**

In order to interact with the FHIR servers in any of the environments, you will need to authenticate with a Google account to gain access to the server's environment.

### Request Access

You will do these steps only one time.

1.  In a browser, go to the server URL from the above. Make sure you are signed out of any Google accounts (e.g. `https://ncpi-api-fhir-service-dev.kidsfirstdrc.org`)

2.  Click `Login with Google` and try to sign in with a Google account:

<img src="docs/images/auth0-signin.png" alt="Auth0 Sign-in" height="400" style="box-shadow:0 10px 16px 0 rgba(0,0,0,0.2),0 6px 20px 0 rgba(0,0,0,0.19)" />

3. You will get a page with a `401 Authorization Required` message:

<img src="docs/images/401-not-auth.png" alt="401 Not Auth" width="400" style="box-shadow:0 10px 16px 0 rgba(0,0,0,0.2),0 6px 20px 0 rgba(0,0,0,0.19)" />

4. Send an email to either [Allison Heath](mailto:heathap@chop.edu), [Alex Lubneuski](mailto:lubneuskia@chop.edu), [Natasha Singh](mailto:singhn4@chop.edu) or [Meen Chul Kim](mailto:kimm11@chop.edu), or a Slack message in the NCPI `#fhir-wg` channel with the Google account you used before and request access for the server URL you wish to access

5. You will receive an email with confirmation of the access

### Authenticate to Access Server Environment

You will do this every time your Cookie expires (~1 week).

1. In a browser, go to `<server URL>/metadata` (e.g. `https://ncpi-api-fhir-service-dev.kidsfirstdrc.org/metadata`)

2. If successful, you will see a page from the server that shows the server's `CapabilityStatement`:

<img src="docs/images/smile-cdr-success.png" alt="Success" style="box-shadow:0 10px 16px 0 rgba(0,0,0,0.2),0 6px 20px 0 rgba(0,0,0,0.19)" />

3. Save the cookie from the response/ The cookie can be found in the response's `Set-Cookie` directive for `AWSELBAuthSessionCookie-0`

### ⚠️ No Need for a User Account on Server

Previously, you needed a user account on the server in order to authenticate with it before sending any other HTTP requests to it. This is no longer needed.

Most users will not need to have an account on the server now since the permissions for anonymous HTTP requests allow one to perform any standard FHIR client operation on the server. This means you can create, read, update, delete, and search for FHIR resources on the server.

### Send Requests to Server

Although you don't need an account on the server to authenticate with it, you will still need to do something to tell the server you have been granted access to it. When you send HTTP requests to the server make sure to include your `AWSELBAuthSessionCookie-0=yadayada` cookie text in a `Cookie` header as follows:

```bash
curl -X GET https://ncpi-api-fhir-service-dev.kidsfirstdrc.org/Patient -L --cookie <the cookie>
```

## Development

You can experiment locally with the FHIR Docker Compose stack. The services/apps included in this are:

- Smile CDR FHIR services (See [Endpoints](#Endpoints) below)
- PostgresSQL database for the server
- FHIR Data Dashboard web app

### Spin up the Docker Stack

1. Clone this repository

```bash
$ git clone git@github.com:kids-first/ncpi-api-fhir-service.git
$ cd ncpi-api-fhir-service
```

2. Get access to the Smile CDR image

   - Create a [Docker Hub](https://hub.docker.com/) account if you don't have one
   - Ask [Natasha Singh](mailto:singhn4@chop.edu) or [Meen Chul Kim](mailto:kimm11@chop.edu) for access to the image (hosted in private Docker Hub repo)

   **🚨 Do not distribute the Smile CDR image as it is only for trial use by the internal team!**

3. Set environment variables in a `.env` file (See `server/settings/dev.env` for example)

**Note:**

The `run_local_server.sh` script requires Docker Hub credentials. First it will look for the environment variables `DOCKER_HUB_USERNAME` and `DOCKER_HUB_PW`. If either of these are not set then it will try to source them from the `.env` file.

4. Deploy server and load the [NCPI Project Forge FHIR model](https://github.com/ncpi-fhir/ncpi-model-forge) (deprecated) into server

```bash
# Deploy server
./scripts/run_local_server.sh

# Load model into server (deprecated)
./scripts/load_model.sh
```

You could also run the steps in `run_local_server.sh` manually. It is just a convenience script which does some setup and then runs `docker-compose up -d`.

### Endpoints

#### [FHIR Data Dashboard](https://github.com/ncpi-fhir/ncpi-ui-fhir-data-dashboard)

- A data browser app intended to give users a quick overview of the data in the FHIR server along with the ability to filter FHIR resources and drill down to view specific resources
- {ROOT_URL}:3000

#### [FHIR API](https://smilecdr.com/docs/tutorial_and_tour/fhir_crud_operations.html)

- The main endpoint ingest developers will use to CRUD FHIR resources
- {ROOT_URL}:8000

#### [FHIR Client Web App](https://smilecdr.com/docs/fhir_repository/fhirweb_console.html)

- A web application used to CRUD FHIR resources for those who do not want to write code
- {ROOT_URL}:8001

#### [Smile CDR Admin API](https://smilecdr.com/docs/fhir_repository/fhirweb_console.html)

- The administration endpoint used to change server configuration, user settings, etc.
- {ROOT_URL}:9000

#### [Smile CDR Admin Dashboard](https://smilecdr.com/docs/modules/web_admin_console.html)

- The administration dashboard which is essentially a frontend to the admin API
- {ROOT_URL}:9100

### Start/Stop the Stack

```bash
# Stop all services
docker-compose stop

# Start all services
docker-compose start
```

### View Service Logs

Once the services are running you can view logs from all services:

```bash
docker-compose logs -f
```

### Reload NCPI FHIR Model (Deprecated)

Run the loader script to reload (DELETE and POST) the NCPI FHIR conformance resources.

The conformance resources are sourced from the `ncpi-model-forge` git repository. The default branch that is used for loading is `master`, but you can supply a different branch if you want.

This script will always do a `git pull` to update the branch before loading the resources. This might result in merge errors if you have made changes to the local branch which you will need to resolve (or do a complete wipe out `git reset --hard origin/<branch>`)

```bash
./scripts/load_model.sh some-other-branch
```

### Server Settings

#### Properties File

Server settings are controlled by modifying the `server/settings/master.properties` Java properties file. The property strings in the file represent a hierarchical structure of config modules in the Smile CDR. Each module pertains to a logical set of functionality (e.g. persistence) in the server. Read more about the config modules [here](https://smilecdr.com/docs/json_admin_endpoints/module_config_endpoint.html).

#### Environment Variable Substitution

Property values in the properties file may also be passed in from the environment using the [environment variable substitution expressions](https://smilecdr.com/docs/installation/installing_smile_cdr.html#variable-substitution). This is useful for passing in secrets like DB credentials.

**Note:**

Any settings changed and saved via the Smile CDR Admin Dashboard or the Smile CDR Admin API will be discarded every time the server is re-deployed. This is because the source of truth for the settings is the properties file and NOT the database. However, this behavior can be changed with the [node.propertysource property](https://smilecdr.com/docs/installation/installing_smile_cdr.html#module-property-source).


CHANGELOG.md
# NCPI FHIR Service Change History

## Release 0.4.0

### Summary

- Emojis: ⬆️ x1, 🔧 x1
- Categories: Ops x1, Other Changes x1

### New features and changes

- [#30](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/30) - ⬆️  Upgrade to 2021.02.R05 - [6365d9f1](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/6365d9f1930d04a53b1cb8c5e051848b58c0f383) by [liberaliscomputing](https://github.com/liberaliscomputing)
- [#29](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/29) - 🔧 Added additional ssl - [1fcb0e79](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/1fcb0e7932d6514aaa5b26368b448e0138e93511) by [alubneuski](https://github.com/alubneuski)


## Release 0.3.1

### Summary

- Emojis: 🐛 x1
- Categories: Fixes x1

### New features and changes

- [#24](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/24) - 🐛 Fix img display on README - [9c583dde](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/9c583dde4e3aec6ed975f784a12a63fd4ad6dd33) by [znatty22](https://github.com/znatty22)


## Release 0.3.0

### Summary

- Emojis: 📝 x1, 🐳 x1, ✨ x1
- Categories: Documentation x1, Ops x1, Additions x1

### New features and changes

- [#22](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/22) - 📝 Improve user access - [c7a65f06](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/c7a65f06125c703753ff9ff65840446991bfc4ab) by [znatty22](https://github.com/znatty22)
- [#20](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/20) - 🐳 Point to NCPI docker img not KF img - [daf8234b](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/daf8234bd267698e73b39a94dcf1d7d267b53d8e) by [znatty22](https://github.com/znatty22)
- [#18](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/18) - ✨ Grant anonymous user FHIR client superuser - [6fcded9f](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/6fcded9fa2d395df006606119720e2b9b3ff2633) by [znatty22](https://github.com/znatty22)


## Release 0.2.0

### New features and changes

- [#15](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/15) -  Added PRD environment in Jenkinsfile - [a2c9f395](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/a2c9f395a44f72ad6728794f18ceb5c627b0b872) by [alubneuski](https://github.com/alubneuski)


## Release 0.1.0

### Summary

- Emojis: ✨ x1, 📝 x4, ⬆️ x1, 🐛 x1, 🔧 x1, 👷 x1, ? x1
- Categories: Additions x1, Documentation x4, Ops x2, Fixes x1, Other Changes x2

### New features and changes

- [#13](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/13) - ✨ Updates from upstream 0.2.0 - [53b13d13](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/53b13d13d87608cb5aa84dcc806d6f24cec5fdd8) by [znatty22](https://github.com/znatty22)
- [#9](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/9) - 📝 Clarify request header instructions - [f4764627](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/f4764627e0c79ca646c28f788bda92bd1f7e2bed) by [fiendish](https://github.com/fiendish)
- [#8](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/8) - ⬆️ Update from upstream - [5ca90959](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/5ca90959374b4c6e261401031115c2ced3e049e9) by [znatty22](https://github.com/znatty22)
- [#7](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/7) - 📝 Update README with specific Cookie - [e3ff7fa4](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/e3ff7fa4ee334f2707a7a9cc458f76460f0e5bfc) by [znatty22](https://github.com/znatty22)
- [#6](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/6) - 📝 Clarify FHIR server authentication - [b99223d1](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/b99223d1041bc9d80a399a4c17992852436a7ab2) by [znatty22](https://github.com/znatty22)
- [#5](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/5) - 🐛 Fix load model script - [5ae63257](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/5ae6325708ae9a5e3c90da67b1cedd3e1218c4da) by [znatty22](https://github.com/znatty22)
- [#4](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/4) - 🔧 Minor updates to scripts and README - [424d26ff](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/424d26ff4670ca84b6188f5cb1b347933e0df3f6) by [znatty22](https://github.com/znatty22)
- [#3](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/3) - 📝 Update README with NCPI stuff - [eccc275a](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/eccc275a69be072a7410835d9da0fb25c5545614) by [znatty22](https://github.com/znatty22)
- [#2](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/2) - 👷 Updated dev,qa CIDR block to become available to the internet - [952a3587](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/952a3587d6426dee6163da94451a7bc985603377) by [alubneuski](https://github.com/alubneuski)
- [#1](https://github.com/ncpi-fhir/ncpi-api-fhir-service/pull/1) -  Modified Jenkinsfile to support AUTH - [7fd0851d](https://github.com/ncpi-fhir/ncpi-api-fhir-service/commit/7fd0851d22957dd017c3ae670783862c710ed1ed) by [alubneuski](https://github.com/alubneuski)

scripts/load_model.sh
#! /bin/bash

# Usage ./scripts/load_model.sh [ my-git-branch ]

# -- Environment Variables --
# Defaults to values in kf-api-fhir-service/.env
# SMILE_CDR_BASE_URL - url of Smile CDR FHIR server
# SMILE_CDR_USERNAME - username to authenticate with before loading resources into server
# SMILE_CDR_PASSWORD - password to authenticate with before loading resources into server

set -eo pipefail

# Args
GIT_REPO_PATH=ncpi-fhir/ncpi-model-forge
GIT_REPO_BRANCH=${1:-master}

# Vars
if [[ ! -f '.env' ]]; then
    cp 'server/settings/dev.env' '.env'
fi
source .env

GIT_REPO=${GIT_REPO_PATH##*/}

echo "$GIT_REPO"
BASE_URL=${SMILE_CDR_BASE_URL:-http://localhost:8000}
SERVER_UNAME=${SMILE_CDR_USERNAME:-$DB_USERNAME}
SERVER_PW=${SMILE_CDR_PASSWORD:-$DB_PASSWORD}


echo "Loading NCPI model $GIT_REPO_PATH into server ..."

# Delete existing model
rm -rf $GIT_REPO

# Git clone the model
git clone "git@github.com:$GIT_REPO_PATH.git"
cd $GIT_REPO
git checkout "$GIT_REPO_BRANCH"

# Setup venv
if [[ ! -d "$(pwd)/$GIT_REPO/venv" ]]; then
    python3 -m venv venv
    source venv/bin/activate
    pip install --upgrade pip
    pip install -e .
fi
source venv/bin/activate

# Publish model to server
fhirutil publish site_root/input/resources/terminology \
--base_url="$BASE_URL" --username="$SERVER_UNAME" --password="$SERVER_PW"

fhirutil publish site_root/input/resources/extensions \
--base_url="$BASE_URL" --username="$SERVER_UNAME" --password="$SERVER_PW"

fhirutil publish site_root/input/resources/profiles \
--base_url="$BASE_URL" --username="$SERVER_UNAME" --password="$SERVER_PW"

fhirutil publish site_root/input/resources/search \
--base_url="$BASE_URL" --username="$SERVER_UNAME" --password="$SERVER_PW"


scripts/users.py
"""
Upsert Smile CDR users

See https://smilecdr.com/docs/json_admin_endpoints/user_management_endpoint.html
for more details on user management

Read JSON file with Smile CDR user objects
(see server/settings/users.example.json) and do following:

- Try creating user
- If user already exists, update user
"""

import argparse
import os
import json
from requests.auth import HTTPBasicAuth
from dotenv import load_dotenv
import secrets

from utils import send_request


def upsert(user_file, base_url, username=None, password=None, headers=None):
    """
    Upsert the smilecdr user payloads

    :param user_file: Path to JSON file with Smile CDR user objects
    :type user_file: str
    :param base_url: Smile CDR server base url
    :type base_url: str
    :param username: Username of user with permission to create/update users
    :type username: str
    :param password: Password of user with permission to create/update users
    :type password: str
    :param headers: Headers to send in HTTP request
    :type headers: str of the form key1=value1 key2=value2
    """
    print(f'Processing user file: {user_file} ...')

    with open(user_file, 'r') as user_json:
        users = json.load(user_json)

    submitted_users = []
    for user in users:
        print(f"Submitting user:{user['username']}")

        url = f'{base_url}/user-management/{user["nodeId"]}/{user["moduleId"]}'

        user.pop('pid', None)
        user['password'] = secrets.token_urlsafe(16)

        # Create headers dict
        header_dict = {}
        if headers:
            for kvpair in headers.split(' '):
                k, v = kvpair.split('=', 1)
                header_dict[k.strip()] = v.strip()

        auth = None
        if username and password:
            header_dict.pop('Authorization', None)
            auth = HTTPBasicAuth(username, password)

        # Send POST request
        response = send_request(
            'post', url, auth=auth, json=user, headers=header_dict
        )
        resp_content = response_content(response)

        # User exists, send PUT request
        if response.status_code == 400 and (
            'already exists' in response.text or
            'manually create system users' in response.text
        ):
            print(
                f'User {user["username"]} already exists, '
                'try updating user ...'
            )
            # Get user's pid
            response = send_request(
                'get', f'{url}?searchTerm={user["username"]}',
                auth=auth, json=user, headers=header_dict
            )
            if response.status_code != 200:
                raise Exception(
                    f'Failed to update user {user["username"]}!'
                )
            # Update user
            pid = response_content(response)['users'][0]['pid']
            user.pop('password', None)
            response = send_request(
                'put', os.path.join(url, str(pid)),
                auth=auth, json=user, headers=header_dict
            )
            resp_content = response_content(response)

        err = False
        if response.status_code not in {200, 201}:
            err = True
        else:
            try:
                resp_content['password'] = user.get('password')
            except Exception:
                err = True
            else:
                submitted_users.append(resp_content)

        if err:
            raise Exception(
                f'Failed to create or update user {user["username"]}!'
            )

    with open(user_file, 'w') as user_json:
        json.dump(submitted_users, user_json, indent=2, sort_keys=True)


def response_content(response):
    """
    Get response content. Try decoding as JSON first, otherwise return str
    version of content
    """
    try:
        resp_content = response.json()
    except json.decoder.JSONDecodeError:
        resp_content = response.text

    return resp_content


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'user_file', help='Path to a JSON file containing user '
        'payloads expected by smilecdr /user-management endpoint'
    )
    parser.add_argument(
        '--base_url', default='http://localhost:9000',
        help='URL of the FHIR server admin endpoint'
    )
    parser.add_argument(
        '--env_file',
        help='Path to an env file with username (DB_USERNAME) password '
        '(DB_PASSWORD) variables needed to create/update users. '
        'If this is not supplied then search for ``.env` in current working '
        'directory. If no env file is found, then variables will be sourced '
        'from environment.'
    )
    parser.add_argument(
        '--headers',
        help='The headers to add to HTTP requests. '
        'Format: header1=value header2=value'
    )
    args = parser.parse_args()

    env_file = args.env_file
    if not env_file:
        env_file = os.path.join(os.getcwd(), '.env')
    if os.path.isfile(env_file):
        load_dotenv(dotenv_path=env_file)

    upsert(
        args.user_file,
        username=os.getenv('DB_USERNAME'),
        password=os.getenv('DB_PASSWORD'),
        base_url=args.base_url,
        headers=args.headers
    )
